src/rsnn/optim/template_based.py

```
@@ -0,0 +1,296 @@
+logger = setup_logging(__name__, console_level="INFO", file_level="DEBUG")
+
+
+def compute_states(synapses, out_spikes, src_spikes, eps):
+    """Returns:
+    - f_states: firing states at spike times with columns f_index, start
+    - syn_states: synaptic transmission states at synaptic event times
+    - rec_states: refractoriness states at refractory event times
+    - states: all states sorted by (f_index, start)
+    """
+    if eps < 0:
+        raise ValueError("epsilon must be positive.")
```

Ruff (0.12.2)

Avoid specifying long messages outside the exception class

(TRY003)